plutus-chain-index/src/Plutus/ChainIndex/Tx.hs

```diff
@@ -19,44 +25,112 @@ module Plutus.ChainIndex.Tx(
     , citxMintingPolicies
     , citxStakeValidators
     , citxValidators
-    , citxTxId
     ) where
 
-import           Control.Lens (makeLenses)
-import           Data.Aeson   (FromJSON, ToJSON)
-import           Data.Map     (Map)
-import qualified Data.Map     as Map
-import           Data.Set     (Set)
-import qualified Data.Set     as Set
-import           GHC.Generics (Generic)
-import           Ledger       (Datum, DatumHash, MintingPolicy, MintingPolicyHash, OnChainTx (..), Redeemer (..),
-                               RedeemerHash, SlotRange, StakeValidator, StakeValidatorHash, Tx (..), TxId,
-                               TxIn (txInType), TxInType (..), TxOut, TxOutRef (..), Validator, ValidatorHash,
-                               datumHash, mintingPolicyHash, redeemerHash, txId, validatorHash)
+import           Control.Lens              (makeLenses)
+import           Data.Aeson                (FromJSON, ToJSON)
+import           Data.Map                  (Map)
+import qualified Data.Map                  as Map
+import           Data.Set                  (Set)
+import qualified Data.Set                  as Set
+import           Data.Text.Prettyprint.Doc
+import           Data.Tuple                (swap)
+import           GHC.Generics              (Generic)
+import           Ledger                    (Address, Datum, DatumHash, MintingPolicy, MintingPolicyHash, OnChainTx (..),
+                                            Redeemer (..), RedeemerHash, SlotRange, StakeValidator, StakeValidatorHash,
+                                            Tx (..), TxId, TxIn (txInType), TxInType (..), TxOut (txOutAddress),
+                                            TxOutRef (..), Validator, ValidatorHash, datumHash, mintingPolicyHash,
+                                            redeemerHash, txId, validatorHash)
 
-data ChainIndexTx = ChainIndexTx {
-    _citxTxId            :: TxId,
-    _citxInputs          :: Set TxIn,
-    _citxOutputs         :: [TxOut],
-    _citxValidRange      :: !SlotRange,
-    _citxData            :: Map DatumHash Datum,
-    _citxRedeemers       :: Map RedeemerHash Redeemer,
-    _citxMintingPolicies :: Map MintingPolicyHash MintingPolicy,
-    _citxStakeValidators :: Map StakeValidatorHash StakeValidator,
-    _citxValidators      :: Map ValidatorHash Validator
+data ChainIndexTx =
+    ValidChainIndexTx {
+      _citxTxId            :: TxId,
+      -- ^ The id of this valid transaction.
+      _citxInputs          :: Set TxIn,
+      -- ^ The inputs to this valid transaction.
+      _citxOutputs         :: [TxOut],
```

Comment:
I don't understand. I already made the change. There are 2 constructors now: `ValidChainIndexTx` and `InvalidChainIndexTx` where the difference is that the former has the `citxOutputs` field and latter doesn't have it.

Reply:
Did you have something else in mind?

Reply:
Yeah I thought it would be something like `_citxOutputs :: ChainIndexOutputs` with

```haskell
data ChainIndexOutputs =
  InvalidTx -- no outputs
  ValidTx [TxOut]
```

so that `ChainIndexTx` would only be a record

Reply:
ok, sorry if that was unclear, I think I misunderstood what you were suggesting.

Reply:
Done. Can you validate?